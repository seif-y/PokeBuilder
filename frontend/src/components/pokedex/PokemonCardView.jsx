import React from "react";
import { Link } from "react-router-dom";
import styles from "./PokemonCardView.module.css";
import PokeType from "../global/PokeType";

/**
 * A basic Pokemon view which displays the Pokemon's sprite, name, dex number and type.
 * @param id: A number representing the Pokemon's dex number
 * @param name: A lowercase string representing the Pokemon's name
 * @param sprite: A string url to the Pokemon's sprite png
 * @param types: An array of strings representing each type name in lowercase
 */
export default function PokemonCardView({ id, name, sprite, types }) {
    // Some Pokemon sprites have their faces covered by their Dex number in the component,
    // so we want to mirror those select sprites for a nicer aesthetic. Array of their ID nums.
    const pokemonSpritesToMirror = [9, 53, 84, 147];

    let capitalisedName = name.charAt(0).toUpperCase() + name.slice(1);

    return (
        <div className={styles.container}>
            <Link className={styles.link} to={`/pokedex/${name}`}>
                <div className={styles.content}>

                    <img className={styles.sprite}
                         style={pokemonSpritesToMirror.includes(id) ? { transform: "scale(-1, 1)" } : null} src={sprite}
                         alt={`PokÃ©mon #${id}`} />
                    <div className={styles.bgCircle} />
                    <div className={styles.nameContainer}>
                        <span className={styles.name}>{capitalisedName}</span>
                    </div>
                    <div className={styles.types}>
                        <PokeType typeName={types[0]} size="small" />
                        {types[1] ? <PokeType typeName={types[1]} size="small" /> : null}
                    </div>
                    <div className={styles.idContainer}>
                        <div>{`#${id}`}</div>
                    </div>

                </div>
            </Link>
        </div>
    );
}
