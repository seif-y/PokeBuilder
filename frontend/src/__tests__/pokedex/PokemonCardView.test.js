import { render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import { StaticRouter } from "react-router";
import PokemonCardView from "../../components/pokedex/PokemonCardView";

const testMon = {
    name: "muk",
    id: 89,
    sprite: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/89.png",
    types: ["poison"],
    displayName: "Muk",
};

const testMonCard = (
    <StaticRouter>
        <PokemonCardView name={testMon.name} id={testMon.id} sprite={testMon.sprite} types={testMon.types} />
    </StaticRouter>
);

describe("PokemonCardView basic tests", () => {
    beforeEach(() => {
        render(testMonCard);
    });

    test("Pokemon card view contains pokemon name", () => {
        const pokemonNameField = screen.getByText(testMon.displayName);
        expect(pokemonNameField).toBeDefined();
    });

    test("Pokemon card view contains pokemon ID", () => {
        const pokemonIdField = screen.getByText(`#${testMon.id}`);
        expect(pokemonIdField).toBeDefined();
    });

    test("Pokemon Sprite exists", () => {
        const pokemonSprite = screen.getByAltText(`PokÃ©mon #${testMon.id}`);
        expect(pokemonSprite).toBeDefined();
    });
});

test("PokemonCardView snapshot test", () => {
    const tree = renderer.create(testMonCard).toJSON();
    expect(tree).toMatchSnapshot();
});
