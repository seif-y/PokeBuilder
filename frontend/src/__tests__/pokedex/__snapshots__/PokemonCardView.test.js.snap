// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonCardView snapshot test 1`] = `
<div
  className="container"
>
  <a
    className="link"
    href="/pokedex/muk"
    onClick={[Function]}
  >
    <div
      className="content"
    >
      <img
        alt="PokÃ©mon #89"
        className="sprite"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/89.png"
        style={null}
      />
      <div
        className="bgCircle"
      />
      <div
        className="nameContainer"
      >
        <span
          className="name"
        >
          Muk
        </span>
      </div>
      <div
        className="types"
      >
        <div
          className="container"
          onClick={[Function]}
          role="button"
          style={
            Object {
              "background": "linear-gradient(225deg,#ce88cb 0%, #c577c4 20%, #bb68bd 40%, #b260b6 60%, #a458ab 80%, #944f9f 100%)",
              "fontSize": "small",
              "width": "55px",
            }
          }
        >
          POISON
        </div>
      </div>
      <div
        className="idContainer"
      >
        <div>
          #89
        </div>
      </div>
    </div>
  </a>
</div>
`;
